core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -359,22 +368,94 @@ public void run() {
    * Is only called at the beginning of the callback function of {@code endOfCurrentBackoff} in the
    * {@link TransportSet#scheduleBackoff} method.
    */
-  void endBackoff() {
+  // TODO(zhangkun83): remove it once the LBv2 refactor is done.
+  final void endBackoff() {
     synchronized (lock) {
       Preconditions.checkState(backoffStatus != null,
           "Error when calling endBackoff: transport is not in backoff period");
       backoffStatus = null;
     }
   }
 
+  /**
+   * Use the picker to try picking a transport for every pending stream and pending ping, proceed
+   * the stream or ping if the pick is successful, otherwise keep it pending.
+   *
+   * <p>If new pending streams are created while reprocess() is in progress, there is no guarantee
+   * that the pending streams will or will not be processed by this picker.
+   */
+  final void reprocess(LoadBalancer2.SubchannelPicker picker) {
+    ArrayList<PendingStream> toProcess;
+    ArrayList<PendingStream> toRemove = new ArrayList<PendingStream>();
+    synchronized (lock) {
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        return;
+      }
+      toProcess = new ArrayList<PendingStream>(pendingStreams);
+    }
+
+    for (final PendingStream stream : toProcess) {
+      PickResult pickResult = picker.pickSubchannel(
+          stream.callOptions.getAffinity(), stream.headers);
+      final ClientTransport realTransport;
+      Subchannel subchannel = pickResult.getSubchannel();
+      if (subchannel != null) {
+        realTransport = obtainActiveTransport(subchannel);
+      } else {
+        realTransport = null;
+      }
+      if (realTransport != null) {
+        Executor executor = streamCreationExecutor;
+        // createRealStream may be expensive. It will start real streams on the transport. If
+        // there are pending requests, they will be serialized too, which may be expensive. Since
+        // we are now on transport thread, we need to offload the work to an executor.
+        if (stream.callOptions.getExecutor() != null) {
+          executor = stream.callOptions.getExecutor();
+        }
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+              stream.createRealStream(realTransport);
+            }
+          });
+        toRemove.add(stream);
+      } else if (!pickResult.getStatus().isOk() && !stream.callOptions.isWaitForReady()) {
+        stream.setStream(new FailingClientStream(pickResult.getStatus()));
+        toRemove.add(stream);
+      }  // other cases: stay pending
+    }
+
+    synchronized (lock) {
+      if (pendingStreams == null || pendingStreams.isEmpty()) {
+        return;
+      }
+      pendingStreams.removeAll(toRemove);
+      if (pendingStreams.isEmpty()) {
+        pendingStreams = null;
```

Comment:
But the code doesn't ever create a new hashmap. The next stream that needs to be pending will cause a NPE. If you want to re-initialize, you'll need to create the new hashmap here.